### PR TITLE
Fix: testing error

### PR DESCRIPTION
**Pull request by ajbt200128**
TL;DR; we were calling semgrep without a main guard, and as a process, causing some weird exec error see [this SO issue](https://stackoverflow.com/questions/24374288/where-to-put-freeze-support-in-a-python-script)

This PR also fixes the same bug that https://github.com/returntocorp/semgrep/pull/8359 fixes, in the same way, so if we merge this we can close that PR and accompanying issue.

## Test plan

```bash
cd semgrep/cli
pipenv install
pipenv shell
semgrep --test --pro --config=. # in archive.zip provided on linear
```

PR checklist:

- [X] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [X] Tests included or PR comment includes a reproducible test plan
- [X] Documentation is up-to-date
- [X] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [X] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

**Comment by ajbt200128** [comment]
Also closes #8254

File: cli/bin/semgrep [hunk 1]
```diff
@@ -68,25 +68,27 @@ def find_semgrep_core_path():
...
-    # tell users to run pysemgrep instead of semgrep and be sure they'll get the
-    # old behavior.
-    path = find_semgrep_core_path()
-    # If you call semgrep-core as osemgrep, then we get osemgrep behavior,
-    # see src/main/Main.ml
-    sys.argv[0] = "osemgrep"
-    os.execvp(str(path), sys.argv)
-else:
-    os.execvp("pysemgrep", sys.argv)
+if __name__ == "__main__":
```
Comment (aryx) [inline cli/bin/semgrep:80]: Can you add a comment explaining this guard for non Python programmers?
And where do we call cli/bin/semgrep in our codebase? and why we need this guard?

**Review by aryx — APPROVED** [review]

File: cli/bin/semgrep (hunk 1 above)
Comment (ajbt200128) [reply cli/bin/semgrep:80]: I added a comment I think explains it, but idk for sure

Comment (ajbt200128) [reply cli/bin/semgrep:80]: Python is weird, python multiprocessing is even weirder

File: cli/bin/semgrep [hunk 2]
```diff
@@ -68,25 +68,36 @@ def find_semgrep_core_path():
...
+        # a different program name. Morever, in case of big problems, we can always
+        # tell users to run pysemgrep instead of semgrep and be sure they'll get the
+        # old behavior.
+        path = find_semgrep_core_path()
+        # If you call semgrep-core as osemgrep, then we get osemgrep behavior,
+        # see src/main/Main.ml
+        sys.argv[0] = "osemgrep"
+        os.execvp(str(path), sys.argv)
+    else:
+        import semgrep.__main__
```
Comment (aryx) [inline cli/bin/semgrep:102]: Oh I didn't see this change in the original PR ... could we restore the execvp("pysemgrep", ...)?
Why does semgrep --test execute the code in this file anyway?

Comment (aryx) [reply cli/bin/semgrep:102]: And why this was causing an error (I still don't understand the scenario in which we crash and the different code paths).

Comment (ajbt200128) [reply cli/bin/semgrep:102]: No we can't remove thjs change, it's part of the fix for this issue, and for fixing autocomplete. There's some explanation for why it fixes autocomplete in the closed PR linked above. I'm also somewhat confused, but I think the best explanation I can give is that for testing, we were calling the invoke semgrep function from multiprocess. So Semgrep is called -> enters into the above -> forks and runs semgrep cli -> test command called -> we run `invoke_semgrep` a few times in parallel -> Since python has the GIL, it can't run a function in a separate thread, it has to start a whole new interpreter (or something, I'm not 100% sure) -> new process means we run all top level code -> since semgrep is imported, this means the above semgrep top level code is called (this phase is called bootstrapping) -> os.exec is called during bootstrapping, before the function that is this process's main is run, and this replaces the mid bootstrapped process before it's had a chance to run -> python does not like this.

The __name__ check stops this from happening, but I'm not sure why the import semgrep.__main__ is needed too. Just that import part fixes autocomplete too (I believe it's something to do with passing environment variables). 

The above is really just my best guess, and I think it's mostly right, but this bug was weird and fix was weird. I think the main reason is that python just really isn't good at multiprocessing, and there's a lot of foot guns

Comment (aryx) [reply cli/bin/semgrep:102]: Then we need to add your big comment in the script.
Also what is autocomplete? I don't think it's supported by osemgrep anyway, so we could get rid of it.

Comment (aryx) [reply cli/bin/semgrep:102]: I'll add some of your comment in my PR
